src/submodule.c

```diff
@@ -1171,9 +1172,21 @@ int git_submodule_update(git_submodule *sm, int init, git_submodule_update_optio
 		 * update the workdir contents of the subrepository, and set the subrepository's
 		 * head to the new commit.
 		 */
-		if ((error = git_submodule_open(&sub_repo, sm)) < 0 ||
-			(error = git_object_lookup(&target_commit, sub_repo, git_submodule_index_id(sm), GIT_OBJ_COMMIT)) < 0 ||
-			(error = git_checkout_tree(sub_repo, target_commit, &update_options.checkout_opts)) != 0 ||
+		if ((error = git_submodule_open(&sub_repo, sm)) < 0)
+			goto done;
+
+		/**
+		 * Look up the target commit in the submodule.
+		 * If it isn't found then fetch and try again.
+		 */
+		if ((error = git_object_lookup(&target_commit, sub_repo, git_submodule_index_id(sm), GIT_OBJ_COMMIT)) < 0) {
```

Comment:
Is there a specific error code that is returned when the target commit could not be found? Could we be more selective and only attempt to fetch for the `not found` error?